Docs for ICs and time steppers

GiudGiud opened the pull request:
and a missing test for a class
Need to add another one for SolutionTimeAdaptiveTimeStepper (uncovered rn)

GiudGiud committed:
Add missing documentation to ICs
Make scalar constant IC value required
Add testing to  array constant IC
Add error testing to array ICs
refs #18333 #17566 #16248 #15968

GiudGiud committed:
Add missing documentation to time steppers
refs #18333 #17566 #16248 #15968

GiudGiud committed:
Add documentation for other timesteppers
refs #18333 #17566 #16248 #15968
Test failures fix: fixup links

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/853115/) on 13e79d3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19129/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19129/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c97975d8b0d95d4fbae2a7e9a582f628c057fef4`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/853213/) on a936390 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19129/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/853242/) on a936390 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c97975</th>
      <th colspan=4>#19129 a93639</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.72%</td>
      <td>81.72%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>70678</td>
      <td>70679</td>
      <td>+1</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15814</td>
      <td>15814</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19129/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19129/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
I added a run app test for that time stepper. Should make sure that it runs without crashes. I think it would be complicated to actually test if the time stepper s behavior is completely unchanged, unless we do something hacky with forcing a certain runtime (like with sleep statement code for example). Not worth it imo

loganharbour commented:
I saw no added coverage for the object you're talking about (when it was up). @dschwen fixed that up in #18871. I'm pretty sure that addition is unnecessary.

GiudGiud commented:
Ok this explains why. That was fixed in between this PR being made and now. Ok I force pushed to remove the test.

loganharbour commented on framework/doc/content/source/ics/ConstantIC.md:
```diff
@@ -1,33 +1,14 @@
 !syntax description /ICs/ConstantIC
 
-## Overview
+This constant value may be set for the whole variable domain, or restricted to particular blocks or boundaries.
```
```suggestion
Sets a constant initial condition described by parameter [!param](/ICs/ConstantIC/value). It can be restricted to particular blocks and boundaries using the [!param](/ICs/ConstantIC/block) and !param](/ICs/ConstantIC/boundary) parameters, respectively.
```

loganharbour commented on framework/doc/content/source/ics/FunctionIC.md:
```diff
@@ -1,33 +1,14 @@
 !syntax description /ICs/FunctionIC
 
-## Overview
+Setting the initial condition using a function may be done for the whole variable domain, or restricted to particular blocks or boundaries.
```
Sets an initial condition via a [Functions/index.md](Function) described by parameter [!param](/ICs/ConstantIC/function). It can be restricted to particular blocks and boundaries using the [!param](/ICs/ConstantIC/block) and !param](/ICs/ConstantIC/boundary) parameters, respectively.

loganharbour commented on framework/doc/content/source/ics/ScalarConstantIC.md:
```diff
@@ -1,32 +1,5 @@
 # ScalarConstantIC
 
-!alert! construction title=Undocumented Class
```
Forget this one?

GiudGiud replied:
there s like nothing to say, class description and parameters cover everything

loganharbour reviewed (changes requested):
Added more details to the most commonly used ICs

GiudGiud committed:
Apply suggestions from code review

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

Update framework/doc/content/source/ics/FunctionIC.md

Apply suggestions from code review: fixup a few equations for the website, and some param links

Add examples for two timesteppers

Update framework/doc/content/source/timesteppers/AB2PredictorCorrector.md

GiudGiud commented on framework/doc/content/source/ics/FunctionIC.md:
```diff
@@ -1,33 +1,14 @@
 !syntax description /ICs/FunctionIC
 
-## Overview
+Sets an initial condition via a [Function](syntax/Functions/index.md) described by parameter [!param](/ICs/ConstantIC/function). It can be restricted to particular blocks and boundaries using the [!param](/ICs/ConstantIC/block) and !param](/ICs/ConstantIC/boundary) parameters, respectively.
```
```suggestion
Sets an initial condition via a [Function](syntax/Functions/index.md) described by parameter [!param](/ICs/FunctionIC/function). It can be restricted to particular blocks and boundaries using the [!param](/ICs/FunctionIC/block) and !param](/ICs/FunctionIC/boundary) parameters, respectively.
```

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/870299/) on 16e85cc wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7607a3</th>
      <th colspan=4>#19129 16e85c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.83%</td>
      <td>81.83%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>71503</td>
      <td>71506</td>
      <td>+3</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15875</td>
      <td>15873</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19129/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19129/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19129/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/source/ics/ConstantIC.md:
```diff
@@ -1,33 +1,14 @@
 !syntax description /ICs/ConstantIC
 
-## Overview
+Sets a constant initial condition described by parameter [!param](/ICs/ConstantIC/value). It can be restricted to particular blocks and boundaries using the [!param](/ICs/ConstantIC/block) and !param](/ICs/ConstantIC/boundary) parameters, respectively.
```
```suggestion
Sets a constant initial condition described by parameter [!param](/ICs/ConstantIC/value). It can be restricted to particular blocks and boundaries using the [!param](/ICs/ConstantIC/block) and [!param](/ICs/ConstantIC/boundary) parameters, respectively.
```

GiudGiud commented on framework/doc/content/source/ics/FunctionIC.md:
```diff
@@ -1,33 +1,14 @@
 !syntax description /ICs/FunctionIC
 
-## Overview
+Sets an initial condition via a [Function](syntax/Functions/index.md) described by parameter [!param](/ICs/FunctionIC/function). It can be restricted to particular blocks and boundaries using the [!param](/ICs/FunctionIC/block) and !param](/ICs/FunctionIC/boundary) parameters, respectively.
```
```suggestion
Sets an initial condition via a [Function](syntax/Functions/index.md) described by parameter [!param](/ICs/FunctionIC/function). It can be restricted to particular blocks and boundaries using the [!param](/ICs/FunctionIC/block) and [!param](/ICs/FunctionIC/boundary) parameters, respectively.
```

GiudGiud commented on framework/doc/content/source/timesteppers/AB2PredictorCorrector.md:
```diff
@@ -1,33 +1,24 @@
-## Overview
+This time stepper first adds an [AdamsPredictor.md] to the problem. The predictor
+uses previous solutions to compute a predicted solution vector. This prediction is
+then compared using a $L\_infinity$ norm to the solution. If the error is lower than `e_max`,
```
```suggestion
then compared using a $L\_infty$ norm to the solution. If the error is lower than `e_max`,
```

GiudGiud commented on framework/doc/content/source/timesteppers/DT2.md:
```diff
@@ -1,33 +1,39 @@
 
-## Example Input File Syntax
+!equation
+dt^{n+1} = dt^n \left(\dfrac{e_tol}{error} \right)^{1/order}
```
```suggestion
dt^{n+1} = dt^n \left(\dfrac{e_{tol}}{error} \right)^{1/order}
```

GiudGiud commented on framework/doc/content/source/timesteppers/DT2.md:
```diff
@@ -1,33 +1,39 @@
+dt^{n+1} = dt^n \left(\dfrac{e_tol}{error} \right)^{1/order}
 
-!! Describe and include an example of how to use the DT2 object.
+where $e_tol$ is the user set tolerance on the error and $order$ is the accuracy order
```
```suggestion
where $e_{tol}$ is the user set tolerance on the error and $order$ is the accuracy order
```

GiudGiud commented on framework/doc/content/source/timesteppers/DT2.md:
```diff
@@ -1,33 +1,39 @@
 
-## Overview
+!equation
+error = \dfrac{||u_{dt} - u_{2dt/2}||_2}{max(||u_{dt}||_2, ||u_{2dt/2}||_2) dt}
```
```suggestion
error = \dfrac{||u_{dt} - u_{2dt/2}||_{L_2}}{max(||u_{dt}||_{L_2}, ||u_{2dt/2}||_{L_2}) dt}
```

GiudGiud commented:
rebased and cleaned up.
Should be ready if it passes the doc build

loganharbour commented on framework/doc/content/source/timesteppers/AB2PredictorCorrector.md:
```diff
@@ -1,33 +1,24 @@
+then the time step is accepted. If not, then it is reduced.
 
-!! Replace these lines with information regarding the AB2PredictorCorrector object.
+The time step is regularly increased based on the `steps_between_increase` parameter. The
```
!param syntax

GiudGiud replied:
```suggestion
The time step is regularly increased based on the
[!param](/Executioner/TimeStepper/AB2PredictorCorrector/steps_between_increase parameter). The
```

moosebuild commented:
Job [Moose Install Target Tests](https://civet.inl.gov/job/882937/) on eb1ee17 : invalidated by @GiudGiud

fluke?
